### PR TITLE
Use function based configuration for more predictable results

### PR DESCRIPTION
**Pull request by kakkoyun**

### CODE REVIEW

File: components/dex.libsonnet [hunk 1]
```diff
@@ -1,28 +1,90 @@
...
+    },
+    kind: 'Secret',
+    metadata: {
+      labels: dex.config.commonLabels,
+      name: dex.config.name,
+      namespace: dex.config.namespace,
+    },
+  },
+
+  pvc: {
```
Comment (metalmatze) [inline components/dex.libsonnet:48]: in kube-thanos we used volumeClaimTemplate here. It's fine probably cause it was called `pvc` before, just slight inconsistencies, I guess.

Comment (kakkoyun) [reply components/dex.libsonnet:48]: I'm gonna address this one before we merge

File: environments/base/manifests/thanos-query-frontend-deployment.yaml [hunk 2]
```diff
@@ -47,34 +47,14 @@ spec:
...
-            "addresses":
-            - "dnssrv+_client._tcp.observatorium-xyz-thanos-query-frontend-memcached.observatorium.svc"
-            "dns_provider_update_interval": "10s"
-            "max_async_buffer_size": 10000
-            "max_async_concurrency": 20
-            "max_get_multi_batch_size": 0
-            "max_get_multi_concurrency": 100
-            "max_idle_connections": 100
-            "timeout": "500ms"
-          "type": "memcached"
```
Comment (yeya24) [inline environments/base/manifests/thanos-query-frontend-deployment.yaml:77]: This is weird. I don't see `query-frontend` related part updated in this pr, but why the yaml was changed?

Comment (yeya24) [reply environments/base/manifests/thanos-query-frontend-deployment.yaml:77]: Oh, I guess you need to re-generate the manifests since the E2E is failing

Comment (kakkoyun) [reply environments/base/manifests/thanos-query-frontend-deployment.yaml:77]: Maybe a rebase error? What's exactly missing?

Comment (kakkoyun) [reply environments/base/manifests/thanos-query-frontend-deployment.yaml:77]: Aha CI also failing because of this

**Review by yeya24 — APPROVED** [review]
Amazing work!

File: components/jaeger-collector.libsonnet [hunk 3]
```diff
@@ -1,167 +1,219 @@
...
+    apiVersion: 'v1',
+    kind: 'Service',
+    metadata: {
+      name: 'jaeger-collector-headless',
+      namespace: j.namespace,
+      labels: { 'app.kubernetes.io/name': $.jaeger.deployment.metadata.name },
+    },
+    spec: {
+      ports: [
+        { name: 'grpc', targetPort: 14250, port: 14250 },
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:55]: This should reference the config above.

File: components/jaeger-collector.libsonnet [hunk 4]
```diff
@@ -1,167 +1,219 @@
...
+    apiVersion: 'v1',
+    kind: 'Service',
+    metadata: {
+      name: 'jaeger-query',
+      namespace: j.namespace,
+      labels: { 'app.kubernetes.io/name': $.jaeger.deployment.metadata.name },
+    },
+    spec: {
+      ports: [
+        { name: 'query', targetPort: 16686, port: 16686 },
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:72]: Reference the config of ports above

File: components/jaeger-collector.libsonnet [hunk 5]
```diff
@@ -1,167 +1,219 @@
...
+    apiVersion: 'v1',
+    kind: 'Service',
+    metadata: {
+      name: 'jaeger-admin',
+      namespace: j.namespace,
+      labels: { 'app.kubernetes.io/name': $.jaeger.deployment.metadata.name },
+    },
+    spec: {
+      ports: [
+        { name: 'admin', targetPort: 14269, port: 14269 },
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:88]: same

File: components/jaeger-collector.libsonnet [hunk 6]
```diff
@@ -1,167 +1,219 @@
...
+    apiVersion: 'v1',
+    kind: 'Service',
+    metadata: {
+      name: 'jaeger-agent-discovery',
+      namespace: j.namespace,
+      labels: { 'app.kubernetes.io/name': 'jaeger-agent' },
     },
+    spec: {
+      ports: [
+        { name: 'metrics', targetPort: 14271, port: 14271 },
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:104]: same

File: components/jaeger-collector.libsonnet [hunk 7]
```diff
@@ -1,167 +1,219 @@
...
+          targetPort: j.config.ports[name],
+        }
+        for name in std.objectFields(j.config.ports)
+      ],
+      volumeMounts: [
+        { name: 'jaeger-store-data', mountPath: '/var/jaeger/store', readOnly: false },
+      ],
+      livenessProbe: { failureThreshold: 4, periodSeconds: 30, httpGet: {
+        scheme: 'HTTP',
+        port: 14269,
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:154]: reference the port in config above

File: components/jaeger-collector.libsonnet [hunk 8]
```diff
@@ -1,167 +1,219 @@
...
+        path: '/',
+      } },
+      readinessProbe: { failureThreshold: 3, periodSeconds: 30, initialDelaySeconds: 10, httpGet: {
+        scheme: 'HTTP',
+        port: 14269,
```
Comment (metalmatze) [inline components/jaeger-collector.libsonnet:159]: same

File: components/memcached.libsonnet [hunk 9]
```diff
@@ -1,60 +1,88 @@
...
-    exporterImage:: error 'must provide exporter image',
-    replicas:: error 'must provide replicas',
-
-    maxItemSize:: '1m',
-    memoryLimitMb:: 1024,
-    connectionLimit:: 1024,
+// Convert number to k8s "quantity" (ie 1.5Gi -> "1536Mi")
+// as per https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+// Original from https://github.com/grafana/jsonnet-libs/blob/master/memcached/memcached.libsonnet
+local bytesToK8sQuantity(i) =
```
Comment (metalmatze) [inline components/memcached.libsonnet:4]:
```suggestion
local bytesToKubernetesQuantity(i) =
```

File: environments/base/manifests/thanos-query-frontend-serviceMonitor.yaml [hunk 10]
```diff
@@ -1,26 +0,0 @@
-apiVersion: monitoring.coreos.com/v1
```
Comment (metalmatze) [inline environments/base/manifests/thanos-query-frontend-serviceMonitor.yaml:1]: We probably want to keep this file and forgot an `serviceMonitor: true` somewhere? :thinking:

**Review by metalmatze — CHANGES_REQUESTED** [review]

File: environments/base/manifests/thanos-query-frontend-serviceMonitor.yaml (hunk 10 above)
Comment (kakkoyun) [reply environments/base/manifests/thanos-query-frontend-serviceMonitor.yaml:1]: On the contrary, we don't generate them in here. This is the only instance. I removed it to be consistent.

**Pull request merged**